File: Java/p0092/Solution.java
package p0092;

import common.ListNode;

public class Solution {
    public ListNode reverseBetween0(ListNode head, int m, int n) {
        if (head == null || head.next == null)
            return head;

        ListNode dummy = new ListNode(-1);
        dummy.next = head;

        int cnt = 0;
        ListNode p = head, rhead = head, tailhead = null, headtail = dummy, pre = dummy;

        while (p != null) {
            cnt++;
            if (cnt == m) {
                headtail = pre;
                rhead = pre.next;
            }
            if (cnt == n) tailhead = p.next;
            pre = p;
            p = p.next;
        }

        pre = null;
        ListNode cur = rhead;
        while (cur != tailhead) {
            ListNode next = cur.next;
            cur.next = pre;
            pre = cur;
            cur = next;
        }

        headtail.next = pre;
        rhead.next = tailhead;

        return dummy.next;
    }

    public ListNode reverseBetween1(ListNode head, int m, int n) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;

        ListNode pre = dummy;
        for (int i = 0; i < m - 1; i++) {
            pre = pre.next;
        }

        ListNode left = pre.next;
        ListNode right = pre;

        for (int i = 0; i < n - m + 1; i++) {
            right = right.next;
        }

        ListNode suc = right.next;

        pre.next = null;
        right.next = null;

        reverseList(left);

        pre.next = right;
        left.next = suc;

        return dummy.next;
    }

    private void reverseList(ListNode head) {
        ListNode pre = null, cur = head;
        while (cur != null) {
            ListNode next = cur.next;
            cur.next = pre;
            pre = cur;
            cur = next;
        }
    }


    public ListNode reverseBetween(ListNode head, int m, int n) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;

        ListNode pre = dummy;
        for (int i = 0; i < m - 1; i++)
            pre = pre.next;

        ListNode cur = pre.next;
        for (int i = 0; i < n - m; i++) {
            ListNode next = cur.next;
            cur.next = next.next;
            next.next = pre.next;
            pre.next = next;
        }

        return dummy.next;
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        ListNode head = ListNode.stringToListNode("[1, 2, 3, 4, 5]");
        ListNode res = s.reverseBetween(head, 1, 5);
        ListNode.prettyPrintLinkedList(res);
    }
}

File: Cpp/p0092/Solution.cpp
#include <iostream>
#include <algorithm>
#include "../common/listnode.h"

using namespace std;

class Solution {
public:
    ListNode* reverseBetween0(ListNode* head, int left, int right) {
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;

        ListNode* p = dummy;
        for (int i = 0; i < left - 1; i++)
            p = p->next;
        ListNode* firstTail = p;
        for (int i = 0; i < right - left + 1; i++)
            p = p->next;
        ListNode* secondHead = p->next;
        p->next = nullptr;

        ListNode* rHead = firstTail->next;
        firstTail->next = nullptr;

        ListNode* newHead = reverseNode(rHead);
        firstTail->next = newHead;
        rHead->next = secondHead;
        
        return dummy->next;
    }

    ListNode* reverseNode(ListNode* root) {
        ListNode* pre = nullptr;
        ListNode* cur = root;
        ListNode* next;

        while (cur != nullptr) {
            next = cur->next;
            cur->next = pre;
            pre = cur;
            cur = next;
        }

        return pre;
    }


    ListNode* reverseBetween(ListNode* head, int left, int right) {
        ListNode* dummy = new ListNode(-1);
        dummy->next = head;
        ListNode* pre = dummy;

        for (int i = 0; i < left - 1; i++)
            pre = pre->next;
        
        ListNode* cur = pre->next;

        for (int i = 0; i < right - left; i++) {
            ListNode* next = cur->next;
            cur->next = next->next;
            next->next = pre->next;
            pre->next = next;
        }

        return dummy->next;
    }
};


int main(int argc, char const *argv[])
{
    Solution* s = new Solution();
    ListNode* root = stringToListNode("[1, 2]");
    root = s->reverseBetween(root, 1, 2);
    prettyPrintLinkedList(root);

    return 0;
}
